### ISSUE TITLE
fix(reranker): score_batch に sub-batching 追加して unbounded OOM を防ぐ (RC-003)

### ISSUE THREAD

**Issue by thkt** [thkt/rurico#123]
## 背景

`RerankerInner::score_batch` (`src/reranker/mlx.rs:35-92`) は `pairs` の N 個分を **単一の flat tensor** で一度に forward する。`pad_sequences` で `[batch_size × bucket_max_len]` の token matrix を構築し、そのまま GPU に upload するため、N が大きいと容易に OOM crash する。

embed 側は `EmbedderInner::embed_documents_batch_chunked_with_metrics` (`src/embed/mlx.rs:269-273`) で `TOKEN_BUDGET / BUCKET_BOUNDS[bucket_idx]` を上限とした **sub-batching** を実装済み。reranker は同等の guard が無く、両 path の対称性が崩れている。

## 根拠 (現状コード)

```rust
// src/reranker/mlx.rs:35-92
pub(super) fn score_batch(
    &mut self,
    pairs: &[(&str, &str)],
) -> Result<Vec<f32>, RerankerError> {
    let mut all_ids: Vec<Vec<u32>> = Vec::with_capacity(pairs.len());
    let mut all_masks: Vec<Vec<u32>> = Vec::with_capacity(pairs.len());

    for (pair_idx, &(query, doc)) in pairs.iter().enumerate() {
        let encoding = self.tokenizer.encode((query, doc), true)
            .map_err(|e| RerankerError::Tokenizer(e.to_string()))?;
        let mut ids = encoding.get_ids().to_vec();
        let mut mask = encoding.get_attention_mask().to_vec();
        truncate_pair(&mut ids, &mut mask, MAX_SEQ_LEN, pair_idx);
        all_ids.push(ids);
        all_masks.push(mask);
    }

    let raw_max = all_ids.iter().map(Vec::len).max().unwrap_or(0);
    let bucket_idx = assign_bucket(raw_max);
    let (flat_ids, flat_mask, batch_size, max_len) =
        pad_sequences(&all_ids, Some(&all_masks), Some(BUCKET_BOUNDS[bucket_idx]));
    // ↑ 全 N pairs を一度に flat tensor 化、N=10000 で 320MB CPU + 40GB GPU
    ...
}
```

呼び出し元 `Reranker::rerank` (`src/reranker.rs:217`) は `self.score_batch(&pairs)` で全 documents を渡すため、caller 側でも batching されない。

## 影響

| シナリオ | 結果 |
|---|---|
| 100 pairs × bucket=512 | ~200KB CPU + ~80MB GPU、問題なし |
| 1000 pairs × bucket=2048 | ~8MB CPU + ~3GB GPU、デバイスにより限界 |
| **10000 pairs × bucket=8192** | **~320MB CPU + ~40GB GPU、確実に OOM** |
| 任意 N (no upper bound) | caller の責任、type で防げない |

reranker の本番ユースケース (検索結果 100-1000 件の re-ranking) では多くの場合は問題ないが、**bulk re-evaluation** や **large candidate pool** のシナリオで明示的にクラッシュする。embed 側の sub-batching と非対称な API 契約は驚きを生む。

## 関連 findings (audit 2026-05-06)

| ID | severity | file | 内容 |
|---|---|---|---|
| **RES-003** | high | `src/reranker/mlx.rs:35-92` | unbounded matrix allocation (root finding) |
| EFF-004 | medium | `src/reranker/mlx.rs:39-52` | `encoding.get_*().to_vec()` × N pairs = 2N heap allocs |
| PRF-005 | high | `src/reranker/mlx.rs:47-51` | 同上 (performance 観点) |
| PRF-006 | low | `src/reranker/mlx.rs:80` | `flat.iter().map(sigmoid).collect()` Vec alloc per call |
| OPS-009 | medium | `src/reranker/mlx.rs:35-93` | per-call timing/batch-shape instrumentation 不在 (Embedder の PhaseMetrics 対称化) |

## 関連 RC

**RC-003** (architecture/inference-path-asymmetry): refactor (`feat/issue-93-model-lifecycle`) は lifecycle (download/cache/probe) を unified にしたが、forward path の契約は unify していない。`trait InferencePath` のような共有 contract が無く、embed 側の改善 (sub-batching, PhaseMetrics) が reranker に back-port されない構造。

### 1. sub-batching の導入

embed 側のパターンを移植：

\`\`\`rust
// src/reranker/mlx.rs::score_batch
const RERANKER_TOKEN_BUDGET: usize = 256_000; // embed と同等

let raw_max = all_ids.iter().map(Vec::len).max().unwrap_or(0);
let bucket_idx = assign_bucket(raw_max);
let bucket_len = BUCKET_BOUNDS[bucket_idx];
let sub_batch_size = (RERANKER_TOKEN_BUDGET / bucket_len).max(1);

let mut all_scores = Vec::with_capacity(pairs.len());
for sub_chunk in all_ids.chunks(sub_batch_size).zip(all_masks.chunks(sub_batch_size)) {
    let (ids_chunk, masks_chunk) = sub_chunk;
    let (flat_ids, flat_mask, batch_size, max_len) =
        pad_sequences(ids_chunk, Some(masks_chunk), Some(bucket_len));
    // ... forward / sigmoid / extend all_scores
    release_inference_output(output);
}
Ok(all_scores)
\`\`\`

### 2. instrumentation 追加 (OPS-009 解決)

`PhaseMetrics` 相当の構造を reranker 用に追加するか、`tracing::debug!` で per-call の (batch_size, sub_batch_count, bucket_len) を構造化 emit。Embedder と同じ field 名を使うことで subscriber 側で対称的に扱える。

### 3. allocation 改善 (PRF-005/006, EFF-004)

- `encoding.get_ids().to_vec()` の 2N alloc は tokenizer batch API (`encode_batch`) で 1 回にまとめられる可能性あり (要 tokenizer crate API 確認)
- `flat.iter().map(sigmoid).collect()` は single-pair score path では `[f32; 1]` で stack 化可能

## 受け入れ条件

- `RerankerInner::score_batch` が sub-batching で `RERANKER_TOKEN_BUDGET` を超えない範囲で forward する
- `Reranker::score_batch` を 5000 pairs (bucket=128) で呼んで OOM しない E2E test
- `tracing::debug!(batch_size, sub_batch_count, bucket_len, ...)` が per-call で emit
- `cargo test --workspace` 全 pass、clippy/fmt clean
- ベンチ: 既存の small-N scenario で latency regression 無し (50 pairs で ±5%)

## downstream 追従 (amici / yomu / recall / sae)

API 影響なし (内部実装のみ、`score_batch` のシグネチャ不変)。merge 後、`Cargo.toml` の `rurico` rev bump のみ。

- [ ] merge 後、amici / yomu / recall / sae 各リポに rev bump 用追従 issue を作成

## 工数

実装 1h (sub-batching loop) + テスト 30min (大規模入力 E2E + 既存 regression)。レビュー別。

## 参照

- audit snapshot: `~/.claude/workspace/history/audit-2026-05-06-084602.json` (RC-003)
- 対称コード: `src/embed/mlx.rs:269-289` (sub-batching reference impl)
